src/gui/tray/ActivityItemActions.qml

```diff
@@ -0,0 +1,143 @@
+import QtQml 2.15
+import QtQuick 2.15
+import QtQuick.Controls 2.3
+import QtQuick.Layouts 1.2
+import Style 1.0
+
+RowLayout {
+    id: root
+
+    spacing: 20
+
+    property string objectType: ""
+    property variant activityLinks: []
+    property bool displayActions: false
+
+    property color moreActionsButtonColor: "transparent"
+
+    property int maxActionButtons: 0
+
+    property Flickable flickable
+
+    signal triggerAction(int actionIndex)
+
+    Repeater {
+        id: actionsRepeater
+        // a max of maxActionButtons will get dispayed as separate buttons
+        model: transformActivityLinks(root.activityLinks)
+
+        function transformActivityLinks(links) {
+            if (links.length === 0) {
+                return links;
+            }
+
+            function createNewLabelAndImageForLink(link) {
+                const imageAndNewlabelForLink = { imageSource: "", imageSourceHovered: "", label: link.label };
+
+                const verb = String(link.verb);
+                const objectType = String(root.objectType);
+
+                // pick proper icon
+                if (verb === "WEB" && (objectType === "chat" || objectType === "call" || objectType === "room")) {
+                    imageAndNewlabelForLink.imageSource = "image://svgimage-custom-color/reply.svg" + "/" + Style.ncBlue;
+                    imageAndNewlabelForLink.imageSourceHovered = "image://svgimage-custom-color/reply.svg" + "/" + Style.ncTextColor;
+                }
+
+                // pick proper label
+                if (verb === "DELETE") {
+                    imageAndNewlabelForLink.label = qsTr("Mark as read")
+                } else if (verb === "WEB" && objectType !== "room" && (objectType === "chat" || objectType !== "call")) {
+                    imageAndNewlabelForLink.label = qsTr("Reply")
+                }
+
+                return imageAndNewlabelForLink;
+            }
+
+            if (links.length > root.maxActionButtons) {
+                return [(Object.assign({}, links[0], createNewLabelAndImageForLink(links[0])))];
+            }
+
+            var reduceLinks = links.reduce(function(reduced, link, index) {
+                reduced.push(Object.assign({}, link, createNewLabelAndImageForLink(link)));
+                return reduced;
+            }, []);
+
+            return reduceLinks;
+        }
```

Comment:
Is there a reason why this can't be done inside the C++ model? If possible, imperative code and data handling tasks are best kept at a minimum in the QML files

Reply:
@claucambra Well, as I said before, we don't want to change the model and try to fit it to how we want a UI to look. We have all the required information in the model already so we can customize it to make it looks the way we want in the UI.  And yes, it also makes sense that QML must be more declarative, so let's see what @camilasan @mgallien and maybe @CarlSchwan think about this?